src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,224 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bedrock.Framework.Protocols
+{
+    public class MessagePipeReader : PipeReader
+    {
+        private readonly IMessageReader<ReadOnlySequence<byte>> _messageReader;
+        private readonly PipeReader _reader;
+
+        private SequencePosition _examined;
+        private SequencePosition _consumed;
+        private ReadOnlySequence<byte> _message;
+        private bool _isThisCompleted;
+        private bool _isCanceled;
+        private bool _isCompleted;
+        private readonly ConsumableArrayBufferWriter<byte> _backlog = new ConsumableArrayBufferWriter<byte>();
+        private bool _allExamined;
+        private bool _advanced = true;
+
+        public MessagePipeReader(PipeReader reader, IMessageReader<ReadOnlySequence<byte>> messageReader)
+        {
+            _reader = reader ?? throw new ArgumentNullException(nameof(reader));
+            _messageReader = messageReader ?? throw new ArgumentNullException(nameof(messageReader));
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            _allExamined = _message.Slice(examined).IsEmpty;
+
+            var consumedLength = (int)_message.Slice(_message.Start, consumed).Length;
+
+            if (_backlog.UnconsumedWrittenCount > 0)
+            {
+                _backlog.Consume(consumedLength);
+            }
+            else
+            {
+                var unconsumed = _message.Slice(consumed);
+                foreach (var m in unconsumed)
+                {
+                    _backlog.Write(m.Span);
+                }
+            }
+
+            // We cannot advance the underlying reader twice without calling _reader.ReadAsync/TryRead in between
+            // If _isCompleted is true we never call reader.ReadAsync/TryRead
+            // So we cannot call _reader.AdvanceTo, and there is no need to because either the buffer was empty,
+            // or we called AdvanceTo in TryCreateReadResult
+            if (!_isCompleted)
+            {
+                // REVIEW: Use the correct value for examined
+                _reader.AdvanceTo(_consumed, _examined);
+            }
+
+            _advanced = true;
+        }
+
+        public override void CancelPendingRead()
+        {
+            _reader.CancelPendingRead();
+        }
+
+        public override void Complete(Exception exception = null)
+        {
+            _isThisCompleted = true;
+            _reader.Complete();
```

Comment:
We shouldn't complete the underlying reader here. The underlying pipe for example could be an http connection and the message reader could be a chunked response body. We don't want the completion of the chunked body to mean the completion of the http connection.